lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -139,95 +150,275 @@
+        if 'arch' in pkg:
+            try:
+                pkg['arch'] = pkg['arch'].split(':')[2]
+            except IndexError:
+                pass
+
+        if 'automatic' in pkg:
+            pkg['automatic'] = bool(pkg['automatic'])
+
+        if 'category' in pkg:
+            pkg['category'] = pkg['category'].split('/', 1)[0]
+
+        if 'version' in pkg:
+            if ',' in pkg['version']:
+                pkg['version'], pkg['port_epoch'] = pkg['version'].split(',', 1)
+            else:
+                pkg['port_epoch'] = 0
+
+            if '_' in pkg['version']:
+                pkg['version'], pkg['revision'] = pkg['version'].split('_', 1)
+            else:
+                pkg['revision'] = 0
+
+        if 'vital' in pkg:
+            pkg['vital'] = bool(pkg['vital'])
+
+        return pkg
+
+
+class PORTAGE(CLIMgr):
+
+    CLI = 'qlist'
+    atoms = ['category', 'name', 'version', 'ebuild_revision', 'slots', 'prefixes', 'sufixes']
+
+    def list_installed(self):
+        rc, out, err = module.run_command(' '.join([self.cli, '-Iv', '|', 'xargs', '-n', '1024', 'qatom']), use_unsafe_shell=True)
+        if rc != 0 or err:
+            raise RuntimeError("Unable to list packages rc=%s : %s" % (rc, to_native(err)))
+        return out.splitlines()
+
+    def get_package_details(self, package):
+        return dict(zip(self.atoms, package.split()))
+
+
+class PIP(CLIMgr):
+
+    CLI = 'pip'
+
+    def list_installed(self):
+        global module
+        rc, out, err = module.run_command([self.cli, 'list', '-l', '--format=json'])
+        if rc != 0:
+            raise Exception("Unable to list packages rc=%s : %s" % (rc, err))
+        return json.loads(out)
 
-    apt_cache = apt.Cache()
-    installed_packages = {}
-    apt_installed_packages = [pk for pk in apt_cache.keys() if apt_cache[pk].is_installed]
-    for package in apt_installed_packages:
-        ac_pkg = apt_cache[package].installed
-        package_details = dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture, source='apt')
-        if package_details['name'] not in installed_packages:
-            installed_packages[package_details['name']] = [package_details]
-        else:
-            installed_packages[package_details['name']].append(package_details)
-    return installed_packages
+    def get_package_details(self, package):
+        return package
+
+
+class PIP2(PIP):
+    CLI = 'pip2'
+
+
+class PIP3(PIP):
+    CLI = 'pip3'
+
+
+class PIP35(PIP):
+    CLI = 'pip3.5'
+
+
+class PIP36(PIP):
+    CLI = 'pip3.6'
+
+
+class PIP37(PIP):
+    CLI = 'pip3.7'
 
 
-# FIXME: add more listing methods
 def main():
+
+    # get supported pkg managers
+    PKG_MANAGERS = dict([(obj.__name__.lower(), obj) for obj in get_all_subclasses(PkgMgr) if obj not in(CLIMgr, LibMgr)])
+
+    # start work
     global module
```

Comment:
Why not initialize the module inside of PkgMgr? You wouldn't have to deal with globals then.

Reply:
PkgMgr doesn't have anything to do with module.  A global is the best fit for an AnsibleModule's scope.

Reply:
Well, you it has methods referring to it. How about passing an instance via initializer?

Reply:
As I said, PkgMgr has nothing to do with AnsibleModule.  They're two separate entities.  If you think of it conceptually, it also makes sense.  This whole file is "an ansible module".  So any object within the file really should be able to access the AnsibleModule object.  Using a global is a better model of the real world here.